### PR TITLE
Upgrade to Orleans 8

### PR DESCRIPTION
**Pull request by richardkooiman**
Happy user of your Orleans Multitenant package here.

Recently tried to upgrade Orleans from 7 to 8 but unfortunately the latest version (1.1.4) is not compatible with Orleans 8 because it relies on the Orleans specific DI extension methods for keyed/named services in the Orleans Core assembly; which are no longer there since Orleans 8 now uses the .NET 8 native keyed service DI methods instead.

I hope you don't mind me making this contribution to see if we can introduce support for Orleans 8. Feel free to do with this as you wish of course.

Changes:
- Target framework upgraded from net7 to net8;
- Orleans version upgrade from 7 to 8.

Build and tests of the main projects pass. The example projects rely on a yet-to-be-released preview version with these changes.

Notes/questions:
- Orleans 8 ships with a new analyzer to detect/warn for missing 'Alias' attributes on custom types. Not sure if/you want to add these, but I left them out for now. This will or might trigger new build warnings. (Links for context: https://github.com/dotnet/orleans/issues/8345, https://github.com/dotnet/orleans/pull/8643)
- I believe I'm only able to make a PR to your main branch. However, should you accept the proposed changes, maybe it would be better to target a orleans8-preview branch instead to be able to release a preview version?
- Would you consider aligning the version numbers with those of Orleans (which itself now aligns with major version numbers from .NET)? This way it might be easier for consumers of the library to understand which version supports which target framework.

### CODE REVIEW

**Comment by VincentH-Net** [comment]
Thanks for this PR @richardkooiman - I had this update on my todo list for a while :-)

- Adding Alias attributes adds a lot of noise, which I dislike - I'll look into that to see if it makes sense for this repo. What I'd really want to catch the versioning errors that this analyzer aims to prevent is a tool I can call in a build pipeline, that can generate a snapshot of all needed info and can compare it to a previous snapshot. So not add Alias everywhere upfront but only last-minute where needed. 99.99% of the alias attributes will never be needed.

- Having a preview branch & package makes sense - I changed the PR base branch as you suggested.

- Aligning version numbers could help a bit but NuGet is very clear on which target frameworks are supported

To include compatibility with .NET Aspire, I would prefer to update to the latest Orleans packages from [their nightly NuGet feed](https://github.com/dotnet/aspire/blob/e344c7c1556bf7f52163ab6fe3ac94ef0c395835/NuGet.config#L13) for now - would that work for you?

**Comment by richardkooiman** [comment]
> Thanks for this PR @richardkooiman - I had this update on my todo list for a while :-)
> 
> * Adding Alias attributes adds a lot of noise, which I dislike - I'll look into that to see if it makes sense for this repo. What I'd really want to catch the versioning errors that this analyzer aims to prevent is a tool I can call in a build pipeline, that can generate a snapshot of all needed info and can compare it to a previous snapshot. So not add Alias everywhere upfront but only last-minute where needed. 99.99% of the alias attributes will never be needed.
> * Having a preview branch & package makes sense - I changed the PR base branch as you suggested.
> * Aligning version numbers could help a bit but NuGet is very clear on which target frameworks are supported
> 
> To include compatibility with .NET Aspire, I would prefer to update to the latest Orleans packages from [their nightly NuGet feed](https://github.com/dotnet/aspire/blob/e344c7c1556bf7f52163ab6fe3ac94ef0c395835/NuGet.config#L13) for now - would that work for you?

You're welcome! :)

- Alias - Yeah I had the same feeling (looks like noise) when looking at it, even though it surely has a purpose.
- Preview branch - Thanks.
- Version numbers - Noted.

If this library would depend on 'Orleans >= 8' people can update Orleans to any 8.x (including nightly/preview) versions right (if let's say needed to support Aspire)? Or what would the result be of the library specifically targeting a nightly version?

**Comment by VincentH-Net** [comment]
> If this library would depend on 'Orleans >= 8' people can update Orleans to any 8.x

@richardkooiman I'll verify 8.0 and 8.1 (which has some changes from 8.0 due to configuration de-inventing, which is leveraged by .NET Aspire support). Probably I will make the examples use 8.1 nightly and the NuGet >= 8 as you suggest. I will do a review shortly btw - busy busy ;-)

**Comment by richardkooiman** [comment]
> > If this library would depend on 'Orleans >= 8' people can update Orleans to any 8.x
> 
> @richardkooiman I'll verify 8.0 and 8.1 (which has some changes from 8.0 due to configuration de-inventing, which is leveraged by .NET Aspire support). Probably I will make the examples use 8.1 nightly and the NuGet >= 8 as you suggest. I will do a review shortly btw - busy busy ;-)

@VincentH-Net Noted. :) Looking forward to it.

**Comment by richardkooiman** [comment]
@VincentH-Net Any idea when you'll have a chance to look into this? :) (No pressure or rush, just curious.)

**Comment by VincentH-Net** [comment]
> Any idea when you'll have a chance to look into this? :) (No pressure or rush, just curious.)

@richardkooiman you have excellent timing :-) I need to update this library for a customer, and I got green light today to do that as part of my paid work (and contribute it back) - will start this Thursday

File: src/Orleans.Multitenant/Internal/Extensions.cs [hunk 1]
```diff
@@ -22,13 +22,13 @@ static class SiloBuilderExtensions
...
-            services.TryAddSingleton<IGrainStorage>(sp => sp.GetServiceByName<MultitenantStorage>(ProviderConstants.DEFAULT_STORAGE_PROVIDER_NAME));
-            services.TryAddSingleton(sp => sp.GetServiceByName<ITenantGrainStorageFactory>(ProviderConstants.DEFAULT_STORAGE_PROVIDER_NAME));
+            services.TryAddSingleton<IGrainStorage>(sp => sp.GetKeyedService<MultitenantStorage>(ProviderConstants.DEFAULT_STORAGE_PROVIDER_NAME));
+            services.TryAddSingleton(sp => sp.GetKeyedService<ITenantGrainStorageFactory>(ProviderConstants.DEFAULT_STORAGE_PROVIDER_NAME));
         }
-        return services.AddSingletonNamedService(name, MultitenantStorageFactory.Create)
-                       .AddSingletonNamedService<IGrainStorage>(name, (s, n) => s.GetRequiredServiceByName<MultitenantStorage>(n))
-                       .AddSingletonNamedService(name, (s, n) => (ILifecycleParticipant<ISiloLifecycle>)s.GetRequiredServiceByName<IGrainStorage>(n))
-                       .AddSingletonNamedService(name, factory);
+        return services.AddKeyedSingleton(name, (s, _) => MultitenantStorageFactory.Create(s, name))
```
Comment (VincentH-Net) [inline src/Orleans.Multitenant/Internal/Extensions.cs:28]: Why the change from 
`MultitenantStorageFactory.Create` to 
`(s, _) => MultitenantStorageFactory.Create(s, name))` ?

I would expect to trust `AddKeyedSingleton` to pass `name` to `Create` - any reason you want to bypass that?

Same for `(s, _) => factory(s, name)`

**Review by VincentH-Net — APPROVED** [review]
Looks all good to me.

I have one question (see my comment on Extensions.cs) but it's OK to merge as-is
I'm still interested in your answer though.

Thanks for this clean PR!

File: src/Orleans.Multitenant/Internal/Extensions.cs (hunk 1 above)
Comment (richardkooiman) [reply src/Orleans.Multitenant/Internal/Extensions.cs:28]: You're right that can be trusted. The "issue" was that the new keyed services methods have their 'key'-argument typed as `object?` while the factories expect a `string`.
So the options were:
1 Do a cast (which very theoretically could fail so might want to catch and explain what went wrong);
2 Change the factories (and as a result the options validator) to receive `object?` instead (would slightly increase the change, but arguably would be an improvement);
3 Using the original name `string` parameter to keep the change the smallest.

Went for option 3, but please feel free to consider another!

**Comment by richardkooiman** [comment]
> Looks all good to me.
> 
> I have one question (see my comment on Extensions.cs) but it's OK to merge as-is I'm still interested in your answer though.
> 
> Thanks for this clean PR!

You're welcome. :)

Thanks for checking and merging the PR! Just replied to your comment.

Any chance we could get a preview release out with these changes or are you planning to add some more before doing that?

**Comment by VincentH-Net** [comment]
> Any chance we could get a preview release out with these changes or are you planning to add some more before doing that?

@richardkooiman I'm fixing the new code analysis warnings - due to `<AnalysisLevel>preview-All</AnalysisLevel>` a lot have popped up and I only release with `<TreatWarningsAsErrors>true</TreatWarningsAsErrors>`. I'm establishing guidelines to which of the new rules apply where + documenting rationale as I go.

I can release a package right after that if you are in a hurry - otherwise I will see if there is more low hanging fruit and release a bit later (no more than a week). Lmk.

**Comment by VincentH-Net** [comment]
@richardkooiman 2.0.0-preview.1 package is [here](https://www.nuget.org/packages/Orleans.Multitenant/2.0.0-preview.1)! - all tests green plus example works.

**Comment by richardkooiman** [comment]
@VincentH-Net Sorry for the late reply - wasn't working a lot due to easter. ;) Thanks so much for the preview release!

**Comment by VincentH-Net** [comment]
> Sorry for the late reply - wasn't working a lot due to easter. ;) Thanks so much for the preview release!

@richardkooiman Did you have a chance to test it out? I'm planning on releasing a non-preview version next week if there are no issues.

**Comment by richardkooiman** [comment]
> > Sorry for the late reply - wasn't working a lot due to easter. ;) Thanks so much for the preview release!
> 
> @richardkooiman Did you have a chance to test it out? I'm planning on releasing a non-preview version next week if there are no issues.

Yes - so far all seems to be working as expected. :) Thanks!
